community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/AwaitIndexProcedure.java

```diff
@@ -0,0 +1,157 @@
+
+import static java.lang.String.format;
+
+public class AwaitIndexProcedure
+{
+    private final ReadOperations operations;
+
+    public AwaitIndexProcedure( KernelTransaction tx )
+    {
+        operations = tx.acquireStatement().readOperations();
+    }
+
+    public void execute( String labelName, String propertyKeyName, long timeout, TimeUnit timeoutUnits )
+            throws ProcedureException
+    {
+        int labelId = getLabelId( labelName );
+        int propertyKeyId = getPropertyKeyId( propertyKeyName );
+        String indexDescription = formatIndex( labelName, propertyKeyName );
+        IndexDescriptor index = getIndex( labelId, propertyKeyId, indexDescription );
+        waitUntilOnline( index, indexDescription, timeout, timeoutUnits );
+    }
+
+    private int getLabelId( String labelName ) throws ProcedureException
+    {
+        int labelId = operations.labelGetForName( labelName );
+        if ( labelId == ReadOperations.NO_SUCH_LABEL )
+        {
+            throw new ProcedureException( Status.Schema.LabelAccessFailed, "No such label %s", labelName );
+        }
+        return labelId;
+    }
+
+    private int getPropertyKeyId( String propertyKeyName ) throws ProcedureException
+    {
+        int propertyKeyId = operations.propertyKeyGetForName( propertyKeyName );
+        if ( propertyKeyId == ReadOperations.NO_SUCH_PROPERTY_KEY )
+        {
+            throw new ProcedureException( Status.Schema.PropertyKeyAccessFailed,
+                    "No such property key %s", propertyKeyName );
+        }
+        return propertyKeyId;
+    }
+
+    private IndexDescriptor getIndex( int labelId, int propertyKeyId, String indexDescription ) throws
+            ProcedureException
+    {
+        try
+        {
+            return operations.indexGetForLabelAndPropertyKey( labelId, propertyKeyId );
+        }
+        catch ( SchemaRuleNotFoundException e )
+        {
+            throw new ProcedureException( Status.Schema.IndexNotFound, e, "No index on %s", indexDescription );
+        }
+    }
+
+    private void waitUntilOnline( IndexDescriptor index, String indexDescription, long timeout, TimeUnit timeoutUnits )
+            throws ProcedureException
+    {
+        try
+        {
+            Predicates.await( () -> isOnline( indexDescription, index ), timeout, timeoutUnits );
+        }
+        catch ( TunnellingException e )
+        {
+            throw e.wrapped;
+        }
+        catch ( TimeoutException e )
+        {
+            throw new ProcedureException( Status.Procedure.ProcedureCallFailed,
+                    "Index on %s did not come online within %s %s", indexDescription, timeout, timeoutUnits );
+        }
+        catch ( InterruptedException e )
+        {
+            throw new ProcedureException( Status.Procedure.ProcedureCallFailed,
+                    "Interrupted waiting for index on %s to come online", indexDescription );
+        }
```

Comment:
These two situations are surfaced with only a variation in text. Do we have "timeout" and/or "interrupted" statuses that we could use instead of a general failure status?